File: nitro-python-example/enclave/enclaveServer.py
import base64
import json
import os
import socket
import hashlib
from Crypto import Random
from Crypto.Cipher import AES
from eth_account import Account
import secrets
from web3.auto import w3

from kms import nitroKms

class nitroServer:

    def __init__(self, region):
        self.__region = region
        pass

    def generateAccount(self, credential, keyId):
        nitro_kms = nitroKms()

        # generate eth private key and calculate the account address
        priv = secrets.token_hex(32)
        private_key = "0x" + priv
        print ("SAVE BUT DO NOT SHARE THIS:", private_key)
        acct = Account.from_key(private_key)
        print("Address:", acct.address)

        # Generate data key by KMS GenerateDataKey API with attestation
        datakeyTextBase64 = nitro_kms.call_kms_generate_datakey(credential, keyId)  
        
        datakey_split = datakeyTextBase64.split("\n")
        datakeyCipherText = datakey_split[0].split(":")[1].strip()
        datakeyPlaintext = datakey_split[1].split(":")[1].strip()

        print("datakeyCipherText: ", datakeyCipherText)
        print("datakeyPlaintext: ", datakeyPlaintext)

        # Encrypt User Private_Key using datakey from KMS, by client-side AES
        # Input datakey string and plaintext string
        aesclient = AESCipher(datakeyPlaintext)

        #convert bytes[] to string
        encrypted_privatekey = str(aesclient.encrypt(private_key),encoding='utf-8')
        content = {
            'encryptedPrivateKey': encrypted_privatekey,
            'address': acct.address,
            'encryptedDataKey': datakeyCipherText.strip()
        }
        print(content)
        return content

    # return the private key's hash value, not an implement of crypty sign operation
    def sign(self, credential, encryptedPrivateKey, encryptedDataKey, transaction):
        nitro_kms = nitroKms()

        # Decrypt encrypted data_key by KMS Decrypt API with attestation
        # Key metadata included in Ciphertextblob, return bytes 
        datakeyPlainTextBase64 = nitro_kms.call_kms_decrypt(credential,encryptedDataKey)
        print("decrypted datakey",datakeyPlainTextBase64)
        datakeyPlaintextBase64String = datakeyPlainTextBase64.split(":")[1].strip()

        # Decrypt private_key with plain data key
        # Input datakey string and encrypted string
        aesclient = AESCipher(datakeyPlaintextBase64String)
        private_key = aesclient.decrypt(encryptedPrivateKey)

        # sign
        transaction_signed = w3.eth.account.sign_transaction(transaction, private_key)
        response_plaintext = {"transaction_signed": transaction_signed.rawTransaction.hex(),
                                      "transaction_hash": transaction_signed.hash.hex()}
        return response_plaintext


class AESCipher(object):

    def __init__(self, key):
        self.bs = AES.block_size
        self.key = hashlib.sha256(key.encode()).digest()

    def encrypt(self, raw):
        raw = self._pad(raw)
        iv = Random.new().read(AES.block_size)
        cipher = AES.new(self.key, AES.MODE_CBC, iv)
        return base64.b64encode(iv + cipher.encrypt(raw.encode()))

    def decrypt(self, enc):
        enc = base64.b64decode(enc)
        iv = enc[:AES.block_size]
        cipher = AES.new(self.key, AES.MODE_CBC, iv)
        return self._unpad(cipher.decrypt(enc[AES.block_size:])).decode('utf-8')

    def _pad(self, s):
        return s + (self.bs - len(s) % self.bs) * chr(self.bs - len(s) % self.bs)

    @staticmethod
    def _unpad(s):
        return s[:-ord(s[len(s)-1:])]
    

def main():
    print("nitro server started ...")

    # Create a vsock socket object
    s = socket.socket(socket.AF_VSOCK, socket.SOCK_STREAM)
    # Listen for connection from any CID
    cid = socket.VMADDR_CID_ANY
    # The port should match the client running in parent EC2 instance
    port = 5000
    # Bind the socket to CID and port
    s.bind((cid, port))
    # Listen for connection from client
    s.listen()

    # read region from environment variable
    region = os.getenv("REGION")
    # server client which call generateWallet or sign
    server = nitroServer(region)

    while True:
        c, addr = s.accept()

        # Get AWS credential sent from parent instance
        payload = c.recv(4096)
        payload_json = json.loads(payload.decode())
        print("payload json: {}".format(payload_json))

        apiCall = payload_json["apiCall"]

        if apiCall == "generateAccount":
            print("generateWallet request")
            credential = payload_json["credential"]
            keyId = payload_json["keyId"]
            result = server.generateAccount(credential, keyId)
            # send back to parent instance
            c.send(str.encode(json.dumps(result)))
            print("generateWallet finished")

        elif apiCall == "sign":
            print("sign request")
            credential = payload_json["credential"]
            transaction = payload_json["transaction"]
            encryptedPrivateKey = payload_json["encryptedPrivateKey"]
            encryptedDataKey = payload_json['encryptedDataKey']

            signedStr = server.sign(
                credential, encryptedPrivateKey, encryptedDataKey, transaction)
            c.send(str.encode(json.dumps(signedStr)))

            print("sign fihished")
        else:
            print("nothing to do")

        c.close()

if __name__ == '__main__':
    main()

File: nitro-golang-example/enclave/enclaveServer.go
package main

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdsa"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"strings"

	"golang.org/x/sys/unix"

	"github.com/ethereum/go-ethereum/common/hexutil"
	"github.com/ethereum/go-ethereum/crypto"
)

type requestPlayload struct {
	ApiCall               string
	Aws_access_key_id     string
	Aws_secret_access_key string
	Aws_session_token     string
	KeyId                 string // this is for generateAccount
	//this 3 is for sign
	EncryptedPrivateKey string
	EncryptedDataKey    string
	Transaction         string
}

type generateDataKeyResponse struct {
	datakey_plaintext_base64  string
	datakey_ciphertext_base64 string
}

type generateAccountResponse struct {
	EncryptedPrivateKey string
	Address             string
	EncryptedDataKey    string
}

func call_kms_generate_datakey(aws_access_key_id string, aws_secret_access_key string, aws_session_token string, keyId string) generateDataKeyResponse {
	var result generateDataKeyResponse
	cmd := exec.Command(
		"/app/kmstool_enclave_cli",
		"genkey",
		"--region", os.Getenv("REGION"),
		"--proxy-port", "8000",
		"--aws-access-key-id", aws_access_key_id,
		"--aws-secret-access-key", aws_secret_access_key,
		"--aws-session-token", aws_session_token,
		"--key-id", keyId,
		"--key-spec", "AES-256")

	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Run()
	if err != nil {
		fmt.Println("kms generate datakey error:", err)
	}

	datakey_split := strings.Split(out.String(), "\n")
	datakeyCiphertext_base64 := strings.TrimSpace(strings.Split(datakey_split[0], ":")[1])
	datakeyPlaintext_base64 := strings.TrimSpace(strings.Split(datakey_split[1], ":")[1])
	result.datakey_plaintext_base64 = datakeyPlaintext_base64
	result.datakey_ciphertext_base64 = datakeyCiphertext_base64

	return result
}

func call_kms_decrypt(aws_access_key_id string, aws_secret_access_key string, aws_session_token string, ciphertext string) string {
	cmd := exec.Command(
		"/app/kmstool_enclave_cli",
		"decrypt",
		"--region", os.Getenv("REGION"),
		"--proxy-port", "8000",
		"--aws-access-key-id", aws_access_key_id,
		"--aws-secret-access-key", aws_secret_access_key,
		"--aws-session-token", aws_session_token,
		"--ciphertext", ciphertext)

	fmt.Println("datakey:", ciphertext)
	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Run()
	if err != nil {
		log.Fatal("kms call", err)
	}
	result := out.String()
	fmt.Println("decrypt result:", result)
	return result
}

func generateAccount(aws_access_key_id string, aws_secret_access_key string, aws_session_token string, keyId string) generateAccountResponse {
	privateKey, err := crypto.GenerateKey()
	if err != nil {
		log.Fatal(err)
	}

	privateKeyBytes := crypto.FromECDSA(privateKey)
	fmt.Println("SAVE BUT DO NOT SHARE THIS (Private Key):", hexutil.Encode(privateKeyBytes))

	publicKey := privateKey.Public()
	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
	if !ok {
		log.Fatal("cannot assert type: publicKey is not of type *ecdsa.PublicKey")
	}

	publicKeyBytes := crypto.FromECDSAPub(publicKeyECDSA)
	fmt.Println("Public Key:", hexutil.Encode(publicKeyBytes))

	address := crypto.PubkeyToAddress(*publicKeyECDSA).Hex()
	fmt.Println("Address:", address)

	datakeys := call_kms_generate_datakey(aws_access_key_id, aws_secret_access_key, aws_session_token, keyId)
	datakey_plaintext_base64 := datakeys.datakey_plaintext_base64
	datakey_ciphertext_base64 := datakeys.datakey_ciphertext_base64

	datakey_plaintext, _ := base64.StdEncoding.DecodeString(datakey_plaintext_base64)

	encryptedPrivateKey := encrypt([]byte(datakey_plaintext), string(privateKeyBytes))

	response := generateAccountResponse{
		EncryptedPrivateKey: encryptedPrivateKey,
		Address:             address,
		EncryptedDataKey:    datakey_ciphertext_base64,
	}
	return response
}

func sign(aws_access_key_id string, aws_secret_access_key string, aws_session_token string, encryptedDataKey string, encryptedPrivateKey string, transaction string) []byte {
	datakey_plaintext_base64 := call_kms_decrypt(aws_access_key_id, aws_secret_access_key, aws_session_token, encryptedDataKey)
	datakey_plaintext_base64_string := strings.TrimSpace(strings.Split(datakey_plaintext_base64, ":")[1])
	datakey_plaintext, err := base64.StdEncoding.DecodeString(datakey_plaintext_base64_string)
	if err != nil {
		log.Fatal("datakey", err)
	}
	private_key := decrypt(datakey_plaintext, encryptedPrivateKey)
	privateKey, err := crypto.ToECDSA([]byte(private_key))
	if err != nil {
		log.Fatal("privateKey error", err)
	}
	data := []byte(transaction)
	hash := crypto.Keccak256Hash(data)
	signature, err := crypto.Sign(hash.Bytes(), privateKey)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("hex:", hexutil.Encode(signature))
	fmt.Println("byte", signature)

	return signature
}

func encrypt(key []byte, message string) string {
	//Create byte array from the input string
	plainText := []byte(message)

	//Create a new AES cipher using the key
	block, err := aes.NewCipher(key)

	//IF NewCipher failed, exit:
	if err != nil {
		log.Fatal(err)
	}

	//Make the cipher text a byte array of size BlockSize + the length of the message
	cipherText := make([]byte, aes.BlockSize+len(plainText))

	//iv is the ciphertext up to the blocksize (16)
	iv := cipherText[:aes.BlockSize]
	if _, err = io.ReadFull(rand.Reader, iv); err != nil {
		log.Fatal(err)
	}

	//Encrypt the data:
	stream := cipher.NewCFBEncrypter(block, iv)
	stream.XORKeyStream(cipherText[aes.BlockSize:], plainText)

	//Return string encoded in base64
	return base64.RawStdEncoding.EncodeToString(cipherText)
}

func decrypt(key []byte, secure string) string {
	//Remove base64 encoding:
	cipherText, err := base64.RawStdEncoding.DecodeString(secure)

	//IF DecodeString failed, exit:
	if err != nil {
		log.Fatal(err)
	}

	//Create a new AES cipher with the key and encrypted message
	block, err := aes.NewCipher(key)

	//IF NewCipher failed, exit:
	if err != nil {
		log.Fatal(err)
	}

	//IF the length of the cipherText is less than 16 Bytes:
	if len(cipherText) < aes.BlockSize {
		fmt.Println("Ciphertext block size is too short!")
	}

	iv := cipherText[:aes.BlockSize]
	cipherText = cipherText[aes.BlockSize:]

	//Decrypt the message
	stream := cipher.NewCFBDecrypter(block, iv)
	stream.XORKeyStream(cipherText, cipherText)

	return string(cipherText)
}

func main() {
	fmt.Println("Start nitro enclave vsock server...")

	fd, err := unix.Socket(unix.AF_VSOCK, unix.SOCK_STREAM, 0)
	if err != nil {
		log.Fatal(err)
	}
	// Bind socket to cid 16, port 5000.
	sockaddr := &unix.SockaddrVM{
		CID:  unix.VMADDR_CID_ANY,
		Port: 5000,
	}
	err = unix.Bind(fd, sockaddr)
	if err != nil {
		log.Fatal("Bind ", err)
	}
	// Listen for up to 32 incoming connections.
	err = unix.Listen(fd, 32)
	if err != nil {
		log.Fatal("Listen ", err)
	}

	for {
		nfd, fromSockAdde, err := unix.Accept(fd)
		if err != nil {
			log.Fatal("Accept ", err)
		}
		fmt.Println("fromSockAdde: ", fromSockAdde)
		fmt.Println("conn is: ", nfd)

		requestData := make([]byte, 4096)
		var playload requestPlayload

		n, err := unix.Read(nfd, requestData)
		if err != nil {
			log.Fatal("Accept ", err)
		}

		err = json.Unmarshal(requestData[:n], &playload)
		if err != nil {
			fmt.Println(err.Error())
		}

		fmt.Println("apicall:", playload.ApiCall)

		apiCall := playload.ApiCall
		fmt.Println(apiCall)

		if apiCall == "generateAccount" {
			result := generateAccount(playload.Aws_access_key_id, playload.Aws_secret_access_key,
				playload.Aws_session_token, playload.KeyId)

			b, err := json.Marshal(result)
			if err != nil {
				fmt.Println(err)
			}
			//  send back to parent instance
			unix.Write(nfd, b)
			fmt.Println("generateAccount finished")
		} else if apiCall == "sign" {
			fmt.Println("sign request")
			result := sign(playload.Aws_access_key_id, playload.Aws_secret_access_key, playload.Aws_session_token,
				playload.EncryptedDataKey, playload.EncryptedPrivateKey, playload.Transaction)
			fmt.Println("result is:", result)
			unix.Write(nfd, result)
			fmt.Println("sign fihished")
		} else {
			fmt.Println("nothing to do")
		}
		unix.Close(nfd)
	}
}
